FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt

```diff
@@ -764,117 +800,207 @@ object PgpMsg {
+        if (!isEncrypted) {
+          isEncrypted = openPgpMetadata?.isEncrypted ?: false
+        }
+
+        if (openPgpMetadata?.isSigned == true) {
+          signedBlockCount++
+
+          if (openPgpMetadata.invalidInbandSignatures.isNotEmpty()
+            || openPgpMetadata.invalidDetachedSignatures.isNotEmpty()
+          ) {
+            val invalidSignatureFailures = openPgpMetadata.invalidInbandSignatures +
+                openPgpMetadata.invalidDetachedSignatures
+
+            hasBadSignatures = invalidSignatureFailures.any {
+              it.validationException.underlyingException != null
+            }
+
+            hasUnverifiedSignatures = invalidSignatureFailures.any {
+              it.validationException.message?.startsWith(
+                "Missing verification certificate", true
+              ) ?: false
             }
           }
-        }
 
-        block.type.isContentBlockType() || MimeUtils.isPlainImgAtt(block) -> {
-          if (block.error != null) {
-            resultBlocks.add(block)
+          if (verifiedSignatures.isEmpty()) {
+            verifiedSignatures.putAll(openPgpMetadata.verifiedSignatures)
           } else {
-            contentBlocks.add(block)
+            val bufferedKeysIds = verifiedSignatures.keys.map { it.keyId }
+            val iteratedKeysIds = openPgpMetadata.verifiedSignatures.keys.map { it.keyId }
+            if (bufferedKeysIds != iteratedKeysIds) {
+              hasMixedSignatures = true
+              verifiedSignatures.putAll(openPgpMetadata.verifiedSignatures)
+            }
           }
         }
+      }
 
-        block.type != MsgBlock.Type.PLAIN_ATT -> {
-          resultBlocks.add(block)
+      if (block is DecryptedAndOrSignedContentMsgBlock) {
+        for (innerBlock in block.blocks) {
+          if (canBeAddedToCombinedContent(innerBlock)) {
+            contentBlocks.add(innerBlock)
+          } else {
+            resultBlocks.add(innerBlock)
+          }
         }
       }
+
+      if (canBeAddedToCombinedContent(block)) {
+        contentBlocks.add(block)
+      } else {
+        resultBlocks.add(block)
+      }
     }
 
     val fmtRes = prepareFormattedContentBlock(contentBlocks)
     resultBlocks.add(0, fmtRes.contentBlock)
 
+    if (signedBlockCount > 0 && signedBlockCount != sequentialProcessedBlocks.size) {
+      isPartialSigned = true
+    }
+
     return ProcessedMimeMessageResult(
-      isReplyEncrypted = isReplyEncrypted,
       text = fmtRes.text,
-      blocks = resultBlocks
+      blocks = resultBlocks,
+      verificationResult = VerificationResult(
+        isEncrypted = isEncrypted,
+        isSigned = signedBlockCount > 0,
+        hasMixedSignatures = hasMixedSignatures,
+        isPartialSigned = isPartialSigned,
+        hasUnverifiedSignatures = hasUnverifiedSignatures,
+        hasBadSignatures = hasBadSignatures
+      )
     )
   }
 
+  private fun canBeAddedToCombinedContent(block: MsgBlock): Boolean {
+    return when {
```

Comment:
could be just 
`= (block.type.isContentBlockType() || MimeUtils.isPlainImgAtt(block)) && block.error == null`